kernel/src/net/socket/ip/raw/mod.rs

```diff
@@ -0,0 +1,429 @@
+    }
+}
+
+impl FileLike for RawSocket {
+    fn read(&self, writer: &mut VmWriter) -> Result<usize> {
+        // TODO: set correct flags
+        let flags = SendRecvFlags::empty();
+        let (read_len, _) = self.recv(writer, flags)?;
+        Ok(read_len)
+    }
+
+    fn write(&self, reader: &mut VmReader) -> Result<usize> {
+        let remote = self.remote_endpoint().ok_or_else(|| {
+            Error::with_message(
+                Errno::EDESTADDRREQ,
+                "the destination address is not specified",
+            )
+        })?;
+
+        // TODO: Set correct flags
+        let flags = SendRecvFlags::empty();
+
+        // TODO: Block if send buffer is full
+        self.try_send(reader, &remote.addr, flags)
+    }
+
+    fn as_socket(&self) -> Option<&dyn Socket> {
+        Some(self)
+    }
+
+    fn status_flags(&self) -> StatusFlags {
+        // TODO: when we fully support O_ASYNC, return the flag
+        if self.is_nonblocking() {
+            StatusFlags::O_NONBLOCK
+        } else {
+            StatusFlags::empty()
+        }
+    }
+
+    fn set_status_flags(&self, new_flags: StatusFlags) -> Result<()> {
+        if new_flags.contains(StatusFlags::O_NONBLOCK) {
+            self.set_nonblocking(true);
+        } else {
+            self.set_nonblocking(false);
+        }
+        Ok(())
+    }
+
+    fn metadata(&self) -> Metadata {
+        // This is a dummy implementation.
+        // TODO: Add "SockFS" and link `RawSocket` to it.
+        Metadata::new_socket(
+            0,
+            InodeMode::from_bits_truncate(0o140777),
+            aster_block::BLOCK_SIZE,
+        )
+    }
+}
+
+impl Socket for RawSocket {
+    fn bind(&self, socket_addr: SocketAddr) -> Result<()> {
+        let endpoint: IpEndpoint = socket_addr.try_into()?;
+
+        let mut inner = self.inner.write();
+        inner.borrow_result(|owned_inner| {
+            let bound_raw = match owned_inner
+                .bind(&endpoint.addr, RawSocketObserver::new(self.pollee.clone()))
+            {
+                Ok(bound_raw) => bound_raw,
+                Err((err, err_inner)) => {
+                    return (err_inner, Err(err));
+                }
+            };
+            (Inner::Bound(bound_raw), Ok(()))
+        })
+    }
+
+    fn connect(&self, socket_addr: SocketAddr) -> Result<()> {
+        let endpoint = socket_addr.try_into()?;
+
+        self.try_bind_ephemeral(&endpoint)?;
+
+        let mut inner = self.inner.write();
+        let Inner::Bound(bound_raw) = inner.as_mut() else {
+            return_errno_with_message!(Errno::EINVAL, "the socket is not bound")
+        };
+        bound_raw.set_remote_endpoint(&endpoint);
+
+        Ok(())
+    }
+
+    fn addr(&self) -> Result<SocketAddr> {
+        let inner = self.inner.read();
+        match inner.as_ref() {
+            Inner::Unbound(_) => Ok(UNSPECIFIED_LOCAL_ENDPOINT.into()),
+            Inner::Bound(bound_raw) => Ok(bound_raw.local_endpoint().into()),
```

Comment:
This implemetantion is unlikely to be correct because the result "port" should represent the protocol:
```python
Python 3.13.1 (main, Dec  4 2024, 18:05:56) [GCC 14.2.1 20240910] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import socket
>>> s = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_TCP)
>>> s.getsockname()
('0.0.0.0', 6)
>>> socket.IPPROTO_TCP
6
```